src/python/pants/core/target_types.py

```diff
@@ -353,6 +354,58 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
     return BuiltPackage(archive, (BuiltPackageArtifact(output_filename),))
 
 
+# -----------------------------------------------------------------------------------------------
+# `gen_rule` target
+# -----------------------------------------------------------------------------------------------
+
+
+class GenRuleCommandField(StringField):
+    alias = "command"
+    required = True
+    help = (
+        "Shell command to execute.\n\n" "The command is executed as 'bash -c <command>' by default."
+    )
+
+
+class GenRuleOutputsField(StringSequenceField):
+    alias = "outputs"
+    required = True
+    help = (
+        "Specify the shell command output files and directories.\n\n"
+        "Use a trailing slash on directory names."
+    )
+
+
+class GenRuleSources(Sources):
+    alias = "sources"
```

Comment:
I think it makes sense to have a sources field, as a way to put an overly complicated command into a script file instead of having it inline in the BUILD file is a big win. Granted, it could be housed in a `files` target, but it feels a bit redundant to have an entire separate goal, for just one script file.

And seeing as the source field was required to be present any way, in order for code gen to work (see relocate files target) might as well use it.

See tests, for example use of the sources field.

Reply:
Probably better in a `shell_library` (soon to be `shell_source`/`shell_sources`) target, then loaded with `dependencies`. I think it makes sense to get rid of `Sources` for this or make it private.

We don't want two ways of "owning" a .sh file imo.

Reply:
I have no strong preference either way. 

IIUC, I can't get rid of the `sources` field, in order to have codegen work. Is using a `_` prefix on the alias for the sources field marking it private?

Reply:
Exactly